Setup Wizard Refactor / Updates

nucleogenesis opened the pull request:
# Setup Wizard

**Description revised 31/1/2023**

## Overview

- Introduces updated designs; layouts/architecture supporting new styling
- Refactors the XState machine to manage more context and each stage of the wizard's flow
- Introduces a script and commands to allow starting the devserver in "App Mode" (more on this below) (thanks @rtibbles !)
- Does not include Learn Only Device flow -- but it does lay some foundation for the machine, routes, and templates for the files to be used
- Implements all combinations of following flows: "On my own", "App Mode", "Import Facility", "Create Facility"

### TODO, Follow up work

To let this aging PR find its way into `develop` finally, despite not being finished, I will enumerate the remaining tasks to complete the work.

I am continuing on with the following work. The LOD flow will be done ASAP, some of the other work may become issues to be prioritized in our backlog. _Critical functionality will be done ASAP._

- [ ] Learn-only (Subset of Users) device forms and flows
- [ ] Step X of Y -- I started a path on this but not sure my idea will work out
- [ ] The file layout in the plugin could use some TLC to simplify the architecture
- [ ] The Select Facility page should be skipped whenever there is only 1 Facility available on the network

## Reviewer Guidance

### Trying again / restarting

#### In the browser

The last step does not actually **provision** the device until you click Continue so long as you're in dev mode -- you will see the Setup Wizard Debugger 3000 that will show some data to you and allow you to restart the flow. This helped me a lot during testing of the UI altogether.

You can also always clear your browser's storage -- or specifcally, the `localStorage.savedState` data which is updated with the entire state machine on each transition between states.

#### In the database

`kolibri manage deprovision` will wipe your database for you

### "App Mode"

Be sure  you run `kolibri plugin enable kolibri.plugins.app` to enable the App plugin.

Then you will need to run this from source using the following command:

`yarn app-devserver`

This will start the server and should print out a URL like `http://localhost:8000/initialize/somelonghash123` -- you will need to follow that URL in order to use the server in "App Mode".

**App Mode** here will result in the "On My Own" and "New Facility" superuser being created automatically, skipping the user creation form.

### Flows To Try

**Each should be done with and without the App Mode to be sure all works as expected**

- [ ] On My Own - Should not have to fill out a user form
- [ ] Group Learning -> Full Device -> New Facility
- [ ] Group Learning -> Full Device -> Import Facility -> Create new superuser 
- [ ] Group Learning -> Full Device -> Import Facility -> Authenticate as existing superuser from imported facility

nucleogenesis committed:
animated kolibri vue component, styles for final onbaording page

nucleogenesis committed:
rename superuser* component and constants where it is now used for any user

nucleogenesis committed:
begin clarifying setup-wizard frontend/backend API methods

nucleogenesis committed:
add presets, set stage for 'setting up kolibri' to wrap things up cleanly

nucleogenesis committed:
clean up api plumbing between fe/be setup wizard resources

nucleogenesis committed:
Rewrite facility import as state machine (large)

- Adds a new "nextEvent" meta property to allow states to be defined
such that they indicate to the consumer/user of the state machine what
event they should send (ie, should they send the user to the FINAL state
or to some other intermediary state)
- Adds "step" numbers to state meta properties to declaratively indicate
which step the user is on at any time; this may be inflexible if trying
to reuse states
- Adds `get_or_create` classmethod to `Facility` to help clean up some
of the logic being used within the SetupWizard backend API that may or
may not need to create a facility (ie, one may have been imported
already at the point we are in the flow)
- Maps import flow steps to states and therefore to specific routes,
making for nicer route names and (hopefully) a clear mapping of the
machine to the UI
- Uses the wizardService injected into the import component steps to
maintain context across steps
- Updates the SelectSuperAdminuser flow at the end of Import Facility
flow to simplify how the form information is handled

pcenov commented:
Hi @nucleogenesis, could you let me and @radinamatic whether this is ready for testing? I did try the 'On my own' flow but after I finish the process I am not able to sign in with the credentials I've provided in the wizard:

https://user-images.githubusercontent.com/79847249/205679629-3ac94c8e-8ef8-4429-a670-7b454478bd3f.mp4

nucleogenesis commented:
Thanks @pcenov I'll test that out today again to see where the issue might be and ping you again when ready.

nucleogenesis committed:
Set default facility name in on-my-own

nucleogenesis committed:
Remove `data.user`, facility.name -> facility_name

nucleogenesis commented:
@pcenov I've resolve the issue where you cannot login

nucleogenesis committed:
Add comments to, remove logs from wizardMachine

nucleogenesis committed:
Add comments to final step SettingUpKolibri

nucleogenesis committed:
Clean up how the selected language is provisioned

pcenov commented:
Hi @nucleogenesis here are my observations from today's round of testing. Let me know if you need help in replicating. Posting them here as a long comment, but if you wish I can create follow-up issues as well:

| Issue| Screenshot |
:-------------------------:|:-------------------------:
| On My Own - The 'Device' plugin is not available while it should as I cannot import channels | ![2022-12-07_17-00-07](https://user-images.githubusercontent.com/79847249/206226732-c17ac2aa-3102-4168-9088-75db5956c8af.png) |
:-------------------------:|:-------------------------:
| Group Learning -> Full Device -> Import all data from an existing learning facility - Same as above, I am able to import a facility but the 'Device' plugin is not available  | https://user-images.githubusercontent.com/79847249/206229563-71f398f4-97ce-426e-8dbc-8fb1ceb3f6c1.mp4 |
:-------------------------:|:-------------------------:
| Group Learning -> Full Device -> Import all data from an existing learning facility - Cannot login with the new super admin that I've created  | https://user-images.githubusercontent.com/79847249/206228437-d733d14f-5926-4aac-901c-9be510400c6f.mp4 |
:-------------------------:|:-------------------------:
| Group Learning -> Full Device -> Create New Learning Facility - Cannot finish the setup process. I tried setting up both a formal and non-formal facility plus changing the default presets.   | https://user-images.githubusercontent.com/79847249/206231152-63130eef-31f9-42a2-920f-cfc5856c142b.mp4  Logs [logs-GrLrn-FullDevice_CreateNew.zip](https://github.com/learningequality/kolibri/files/10177818/logs-GrLrn-FullDevice_CreateNew.zip) |
:-------------------------:|:-------------------------:

marcellamaki commented on kolibri/plugins/setup_wizard/api_urls.py:
```diff
@@ -1,15 +1,15 @@
 from .api import SetupWizardRestartZeroconf
 
-router = routers.DefaultRouter()
+router = routers.SimpleRouter()
```
is this a Vue upgrade, or a change you decided on? I don't have any context on what the difference, if any, is

nucleogenesis replied:
I think @rtibbles might be better able to explain the difference -- but if I didn't do this, something broke and when I changed it things worked :)

rtibbles replied:
The DefaultRouter does some introspection to generate an API view at the root of where the router is mounted: https://www.django-rest-framework.org/api-guide/routers/#defaultrouter

My assumption is that this introspection probably breaks because of our own tweaks and adjustments to DRF.

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/machines/wizardMachine.js:
```diff
@@ -126,13 +143,16 @@ export const wizardMachine = createMachine({
           cond: isGroupSetup,
           target: 'deviceName',
         },
+        // There is no fallback path here; if neither `cond` above is truthy, this will break
```
This seems both correct, and also, potentially tricky. Are there other places where we are using a state machine that we have handled a similar scenario?

nucleogenesis replied:
Not that I know of -- ideally we would have some kind of error state for the Setup Wizard that we could put people into with a "start over" button and some information they can send us to help debug? This would require some design and new strings but could be a worthwhile thing to add? (cc @jtamiace)

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/SetupWizardIndex.vue:
```diff
@@ -114,6 +114,8 @@
 
       this.service.onTransition(state => {
+        console.log('ontransition', state);
+        console.log(this.service);
```
logs

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/importFacility/ImportAuthentication.vue:
```diff
@@ -23,21 +23,23 @@
+    },
+    mounted() {
+      console.log('ImportAuthentication', this.wizardService.state);
+    },
```
log

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/importFacility/SelectFacilityForm.vue:
```diff
@@ -75,24 +67,30 @@
       },
     },
+    mounted() {
+      console.log('ImportAuthentication', this.wizardService.state);
```
and here. now I'm wondering if there is a particular reason these are being logged? if so - sorry :D

nucleogenesis replied:
Nope I just didn't do a log audit before making the PR :sweat:

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/importFacility/SelectFacilityForm.vue:
```diff
@@ -104,6 +102,15 @@
+      handleContinue() {
+        //$emit('click_next', { facility: selectedFacility, device: device })
+        console.log(this.wizardService._state);
+        console.log(this.wizardService.state);
```
some cleanup

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/onboarding-forms/DefaultLanguageForm.vue:
```diff
@@ -24,8 +24,6 @@
       handleSubmit() {
-        const currentLanguageId = this.$store.state.onboardingData.language_id;
-        this.$store.commit('SET_LANGUAGE', currentLanguageId);
         this.wizardService.send('CONTINUE');
```
can you remind me where the language setting is handled now? is it just in the pop up that is outside the scope of the flow?

nucleogenesis replied:
The `LanguageSwitcherList` makes the call to the backend to set the default language. This was only stored in Vuex for the frontend's reference previously I believe

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/onboarding-forms/PersonalDataConsentForm.vue:
```diff
@@ -48,7 +48,24 @@
+          return;
+        }
+        // TODO Add an Error State with a "Start over" button? Something better than
+        // "this silently fails" if something goes wrong for the user
```
Yes, I was wondering about this too - and I think maybe we chatted about it in a pair programming session. not a blocker, but would possibly be good to discuss before release

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/onboarding-forms/SettingUpKolibri.vue:
```diff
@@ -1,50 +1,133 @@
+          // 2.2) Import Facility
+          // We already have the facility imported, just provision and redirect
+          this.provisionDevice();
+        }
```
not at all blocking - I think having some comments here is great, but maybe you don't need all of them? I think the ones that are Booleans of a rather descriptive computed prop are okay to skip, personally, and doing so makes it a bit easier to actually look at the _code_ . But I think it other places like 

> // We already have the facility imported, just provision and redirect

where the logic is something you have to have kept in your brain as you were reading, it's actually really helpful.

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/onboarding-forms/SettingUpKolibri.vue:
```diff
@@ -1,50 +1,133 @@
+        SetupWizardResource.provisiondevice(this.deviceProvisioningData)
+          .then(() => {
+            // FIXME In dev mode, we'll wait 5 seconds before moving along so we can see the page
+            // ... maybe we keep this?
```
I think waiting a few seconds might be nice, even if it is mostly like a "foaming toothpaste" effect. Maybe there is some UX research about this? sometimes realllyy short wait times feel glitchy to me.

nucleogenesis replied:
Thoughts @jtamiace ?

nucleogenesis replied:
And @tomiwaoLE ?

jtamiace replied:
What part of the onboarding flow is this? Having a hard time telling

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/onboarding-forms/UserCredentialsForm.vue:
```diff
@@ -13,7 +13,12 @@
     -->
     <slot name="form">
+      <!-- Hiding the fullname and username textboxes, but their values are filled in and presumed
+           valid if we're given the user that we're taking credentials for (ie, just entering
```
> presumed valid

if there is an error, what happens? do we get the validation errors and the form with fill-in textboxes again?

nucleogenesis replied:
Basically, IMO, this can use some refactoring because it basically uses one form in two use cases:
- Get a password for an imported user (ie, we imported a facility and they want to use an existing super admin account on that faciliy, so we get the password from them)
- Creating a whole new superuser

The convenient thing is that in the end the `SettingUpKolibri` component gets the same value (namely `password`) from the same place -- but it might just be more clear if we separate those two needs into separate components and data storage (and/or state machine context?)

*But to answer the actual question...*

So the assumption here is that `username` and `fullname` are valid primarily because they are values that we got from another Kolibri in this particular case where we hide the form elements.

marcellamaki commented on kolibri/plugins/setup_wizard/assets/src/views/onboarding-forms/UserCredentialsForm.vue:
```diff
@@ -91,63 +96,78 @@
+          if (this.$refs.password && this.$refs.password.isValid) {
+            // In this case, we have been given a separate password form in our #form slot
+            // and it's ref can tell us if it's valid and whether we should continue
+            console.log('OKGO');
```
log

marcellamaki reviewed (commented):
Hi @nucleogenesis this is great! I did an initial code read through and left some comments (some were just annoying "console log!") and some are more open questions. 

I don't have any major blockers right now (although, I defer proper review of the python code to...anyone not me). I'm going to do a second read and some manual QA on Monday and will write a follow up review then.

rtibbles commented on kolibri/plugins/setup_wizard/api.py:
```diff
@@ -32,6 +32,141 @@ def has_permission(self, request, view):
         return subset_of_users_device
 
 
+class SetupWizardResource(ViewSet):
```
This appears to be replacing the core provision device endpoint: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/device/api.py#L78

which has decent test coverage of different pathways: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/device/test/test_api.py#L50

I would much rather we add these conditionals to that endpoint and test them, rather than creating a brand new endpoint here.

rtibbles commented on kolibri/core/auth/models.py:
```diff
@@ -1415,6 +1415,23 @@ def add_coaches(self, users):
         self.remove_role(user, role_kinds.COACH)
 
+    @classmethod
+    def get_or_create(cls, facility_name):
```
I am not sure what this class method is giving us over just using `Facility.objects.get_or_create(name=facility_name)`

nucleogenesis replied:
Oh just me not knowing or checking if that's a built-in function

rtibbles commented on kolibri/plugins/setup_wizard/assets/src/views/onboarding-forms/SettingUpKolibri.vue:
```diff
@@ -1,50 +1,133 @@
-      isOnMyOwnSetup() {
-        return this.wizardService.state.context.onMyOwnOrGroup == UsePresets.ON_MY_OWN;
+      createAndProvisionOnMyOwnUserDevice() {
+        SetupWizardResource.createonmyownuser(this.facilityUserData).then(() =>
```
I don't understand why the user creation and provisioning steps have been separated in these three instances, when they are then just chained without any intermediary state? This could simply use the existing endpoint without any issues, and one fewer request.

rtibbles reviewed (commented):
I am confused by the backend recreation here and the separation of the provision and user creation steps.

The provisioning step needs to be separated from import, but not from the flows where new users are created. I would much rather see any tweaks be made in the existing device provisioning endpoint rather than creating entirely novel, untested code paths here.

If any refactor is warranted, it is probably more that the flows that involve user creation might be better done as asynchronous provisioning tasks (which is when the separation of finalized provisioning from all the main setup would be warranted, so as to be able to monitor tasks as an unlogged in user until provision has been finalized).

nucleogenesis committed:
make machine more friendly to visualization, add comments explaining how to do it

nucleogenesis committed:
add debugging screen for dev mode in final stage

nucleogenesis committed:
Fix and improve machine actions, context help

nucleogenesis committed:
Align with changes to state machine

nucleogenesis committed:
Bootstrap properties for device provision data FE

nucleogenesis committed:
Styling to fit all data on debug finish screen

nucleogenesis committed:
Mitigate getting stuck on loading task page

nucleogenesis committed:
null default for superuser in final setup (FE)

nucleogenesis committed:
wire up superuser to provisioning data; fix error error

nucleogenesis committed:
setup routes for lod

nucleogenesis committed:
update the one part of vuex we still use here, fix refs

nucleogenesis committed:
include os_user in device_provisoining data

nucleogenesis committed:
app mode script registered proper fn

nucleogenesis committed:
provision: if facility id given, get it and use it

nucleogenesis committed:
setupwiz debug: add button to proceed after data view

nucleogenesis committed:
sqaushthis

nucleogenesis committed:
onmyown, provisioning complies with device provision api

nucleogenesis committed:
add setSuperuser actions where applicable

nucleogenesis committed:
add check for APP_KEY_COOKIE_NAME before trying to get it

nucleogenesis committed:
use presets in setup facility form

nucleogenesis committed:
lint

nucleogenesis committed:
init lod routes/machine

nucleogenesis committed:
comment out lod components in route (for now)

nucleogenesis committed:
machine name fixes

nucleogenesis committed:
remove timeout from loading page

nucleogenesis committed:
another comment for lod import

pcenov commented:
Hi @nucleogenesis, here's a list with issues I was able to identify in the [latest .deb build asset](https://buildkite.com/learningequality/kolibri-debian/builds/5826):

| Issue| Screenshot |
:-------------------------:|:-------------------------:
| Group learning > Full device > Create a new learning facility - The user entered 'facility name' value is replaced by a ‘Default Facility Name’ value | https://user-images.githubusercontent.com/79847249/216097560-4397e3eb-aa77-4e4b-b425-f5418cd2a49f.mp4 |
:-------------------------:|:-------------------------:
| Group learning > Full device > Create a new learning facility - The already selected/entered data is not displayed when using the ‘Go back’ button | https://user-images.githubusercontent.com/79847249/216099443-9d1e9f8e-9f1b-4436-a273-275baacb6295.mp4 |
:-------------------------:|:-------------------------:
| Group learning > Full device > Create a new learning facility - Non-formal - When setting up a ‘Non-formal facility’ the ‘Yes’ option should be selected as default while currently for both ‘Non-formal’ and ‘Formal’ it is set to ‘No’ | https://user-images.githubusercontent.com/79847249/216097800-cf8dfc6b-d448-4059-bb65-9f3daf76139e.mp4 |
:-------------------------:|:-------------------------:
| Group learning > Full device > The imported facility is duplicated |  https://user-images.githubusercontent.com/79847249/216098375-7baedc6a-b26f-4dc6-a018-bbbe9c5ae301.mp4 |
:-------------------------:|:-------------------------:
| Group learning > Full device > Import all data from an existing learning facility - the ‘Go back’ option is not functional, getting an error in the console |  https://user-images.githubusercontent.com/79847249/216098139-077c3018-b8d7-4306-a8fd-25976ccfaad0.mp4 |
:-------------------------:|:-------------------------:
| Group learning > Full device > Import all data from an existing learning facility - ‘Unknown’ displayed while importing  | ![2023-02-01_18-11-44](https://user-images.githubusercontent.com/79847249/216098834-7fbccf48-74a3-489e-8d96-332c342d4c24.png) |
:-------------------------:|:-------------------------:

nucleogenesis committed:
be sure we provision with given facility name when creating facility

nucleogenesis committed:
default selectedFacility to null; default the facility name in final steps

nucleogenesis committed:
allow event on BACK to be defined, handle BACK by setting data, cleanup

nucleogenesis commented:
@pcenov I have addressed all of the issues noted except the loading issues -- I'm seeing "unknown" as well as it goes up to  5000% before completing so I'll be working on that.

Just a note though that unless this is an error specifically in the code here, I may address it in a separate PR -- will update tomorrow.

I will also look for some live code review from @rtibbles 

From that I will make quick fixes and make issues for anything else so that this can be merged and others can take on some work in the setup wizard to become more familiar.

nucleogenesis committed:
fix password requirement type ambiguity

nucleogenesis committed:
fix tests

pcenov commented:
Hi @nucleogenesis, there are still a few issues in the [latest .deb build asset](https://buildkite.com/learningequality/kolibri-debian/builds/5893#0186198f-10b3-417b-9c2f-09ae37a24ca1) the main one being the facility import option which is not functioning at all now:

| Issue| Screenshot |
:-------------------------:|:-------------------------:
| Group learning > Full device > Create a new learning facility - Now I am getting a 500 server error if  'Enable passwords on learner accounts' is set to 'Yes' | https://user-images.githubusercontent.com/79847249/216993472-f8b5bd79-26ed-4776-9e9b-18578280235b.mp4 |
:-------------------------:|:-------------------------:
| Group learning > Full device > Import all data from an existing learning facility -  Now it's not possible to import a facility at all. I tried several times but I am always getting a 400 bad request error for deviceprovision | https://user-images.githubusercontent.com/79847249/216994946-12554f28-478c-43f8-bf02-d7122e15de8b.mp4 |
:-------------------------:|:-------------------------:
| Group learning > Full device > Import all data from an existing learning facility - the ‘Go back’ option is not working | https://user-images.githubusercontent.com/79847249/216996078-04bafdf6-48d9-47e7-b23a-e98d5bb7ebb2.mp4 |
:-------------------------:|:-------------------------:

marcellamaki commented:
Hi @pcenov - thank you so much for the very thorough testing on all of this work. Can you clarify - are these problems _only_ present on the deb asset, or have you just started you testing beginning with that asset first, and you will follow up with additional updates if these problems also exist in the other assets as well? thank you!

pcenov commented:
Hi @marcellamaki the issues are valid for both the .deb and .exe assets, not OS specific.

nucleogenesis committed:
clean up how provisioning data is gathered

nucleogenesis committed:
update logic for setting up api payload

nucleogenesis commented:
@pcenov -- I found a couple other bugs in addressing your comments and have updated the PR -- should be ready for a retest of the workflows On My Own, Full Facility and Facility Import.

Thanks!

pcenov commented:
Hi @nucleogenesis I've retested the main scenarios for the implemented workflows On My Own, Full Facility and Facility Import and these are the outstanding issues, all of which have been previously reported and can be filed as separate issues:

| Issue| Screenshot |
:-------------------------:|:-------------------------:
| Group Learning -> Full Device -> Import all data from an existing learning facility - As previously reported, I am able to import a facility but the 'Device' plugin is not available | https://user-images.githubusercontent.com/79847249/218780914-f993c43e-7239-4adf-9ed4-11512ae6388b.mp4 |
:-------------------------:|:-------------------------:
| Group learning > Full device > Import all data from an existing learning facility - the ‘Go back’ option is not working correctly | https://user-images.githubusercontent.com/79847249/218781113-14b3be8b-ae74-486f-a640-abf07e446d3a.mp4 |
:-------------------------:|:-------------------------:
| Group learning > Full device > Import all data from an existing learning facility - ‘Unknown’ displayed while importing, incorrect percentage value  | ![2023-02-14_17-27-32](https://user-images.githubusercontent.com/79847249/218782502-a2d52d9f-ac07-4583-b4dc-8b458c6a1f60.png) |
:-------------------------:|:-------------------------:

nucleogenesis commented:
Thank you @pcenov -- I'll merge once this is approved

nucleogenesis commented:
Noting also that each of those will be either addressed in the Learn-only Device follow-up PR or made into a separate follow-up issue

radinamatic reviewed (approved):
It's been a long road but seems like we're almost there, with just a few over-spills 🙂 
Thank you all! 👏🏽 💯 :shipit:
